### DISCUSSION TITLE
Clicking on assembly code on a switch statement of a function sometimes leads to an undefined function on the switch statement instead of staying in the same function

### DISCUSSION THREAD

**Discussion by H-A-M-G-E-R** [NationalSecurityAgency/ghidra#7158] [General]

**Comment by ryanmkurtz** [comment]
Can you please provide a sample binary and and address where this is happening?  We need something more concrete to go off of.

**Comment by H-A-M-G-E-R** [comment]
decompiling a GBA game

**Comment by ghidracadabra** [comment]
It's possible that Ghidra figured out that it should disassemble the bytes corresponding to a case of a switch statement without being able to place the resulting instructions into the body of a function (there are analyzers that look for various bit patterns to guide disassembly).  The slide deck improvingDisassemblyAndDecompilation.pdf (included with the Ghidra distribution) walks through an example of manually fixing a switch statement.

**Comment by H-A-M-G-E-R** [comment]
Example:
![image](https://github.com/user-attachments/assets/398509b1-6109-4d57-8d28-15d876a25aed)
![image](https://github.com/user-attachments/assets/a931cdea-fd8c-4423-bd3c-5e523b5f9cfb)
i never want that to happen

**Comment by H-A-M-G-E-R** [comment]
I think the fix to this is https://github.com/NationalSecurityAgency/ghidra/issues/7206#issuecomment-2524295292: regenerating switch tables.

**Comment by emteere** [comment]
Sometimes the decompilation improves in newer Ghidra versions that allow a switch stmt to recover, where it did not recover in past versions.  Other times it could be because something was resolved that now allows the switch to recover.

If the decompiler is showing the switch correctly and the listing is not, then clearing the instruction and re-creating the switch with analysis should always solve the issue.

Most likely an older version of Ghidra couldn't recover the switch and the table was found by later speculative analysis like the address table analyzer.

When I see a bad switch I usually try to delete the switch instruction and re-disassemble it to diagnose the issue.  If the function body is bad after that you can delete and re-create the function and it should fix up.

If that doesn't correct it, then I'll look for problematic functions, non-returning, side-effects of a function call like affecting the stack, signatures not correct, or other flow issues.

I'll close this for now if it has been fixed for you.
